### PR TITLE
[FLINK-16655][FLINK-16657] Introduce embedded executor and use it for Web Submission

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

The `EmbeddedExecutor` is part of the [FLIP-85](https://cwiki.apache.org/confluence/display/FLINK/FLIP-85+Flink+Application+Mode) effort,  which aims at giving the user the option of running his/her `main` method on the cluster, rather than the client. 

Running the user's main method in "Application Mode" implies:
1) launching a dedicated cluster for the application's jobs
2) running the user's main on the cluster, alongside the Dispatcher

The `EmbeddedExecutor` is an `Executor` which is assumed to run alongside the `Dispatcher` of the cluster, on the same machine.  Conceptually, it is like the existing `Executors` for session clusters, with the difference that this time there is no need to go through REST as this will be running already on the same machine as the Dispatcher.

Given that the Web Submission already runs the user's `main` on the cluster, this PR apart from introducing the `EmbeddedExecutor`, it also wires it to the `JarRunHandler` so that:
1) we can test the executor already
2) we get (partially for now) rid of the pattern of throwing `ProgramAbortException` in order to extract the `JobGraph`. 

## Brief change log

See the description above.
It introduces the `EmbeddedExecutor` with a custom `PipelineExecutorLoader` which is only aware of this executor and is used in the `JarRunHandler`. 

The handler now, instead of extracting the `JobGraph` itself and submitting to the `Dispatcher`, now it executes the user's main method and let's the new executor do the job graph extraction and submission. 

To not block any threads on the web frontend, we launch the job in `detached` mode, which is reasonable, as web submission is expected to be a "fire-and-forget" action, and also it is compatible with the current behaviour which does not return the results of operations like `collect()`, `count()` or `print()`. 

The only difference compared to before is that now, a job with a `print()` will throw an exception (as is the case for detached submission from the CLI), rather than execute the job and silently return nothing.

## Verifying this change

It was tested manually on Yarn and local cluster and also it is tested through the updated tests for the `JarRunHandler`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by TisonKun** [comment]
@flinkbot run azure

File: flink-clients/src/main/java/org/apache/flink/client/deployment/JobSubmitter.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.apache.flink.runtime.jobgraph.JobGraph;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An interface to be implemented by entities that are
+ * allowed/capable to submit {@link JobGraph JobGraphs} for execution.
+ */
+@Internal
+public interface JobSubmitter {
```
Comment (TisonKun) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/JobSubmitter.java]: The only usage of this interface is

```java
public class EmbeddedClient implements JobClient, JobSubmitter {
```

If we don't need this abstraction anywhere else, we should remove it; if we need it, I'll doubt why `ClusterClient` doesn't `extends` it.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		jobGraph.addJars(jars);
+		jobGraph.setClasspaths(classpaths);
+		jobGraph.setSavepointRestoreSettings(savepointRestoreSettings);
+
+		getJobId(configuration).ifPresent(jobGraph::setJobID);
+
+		return jobGraph;
+	}
+
+	private List<URL> decodeUrlList(final Configuration configuration, final ConfigOption<List<String>> configOption) {
```
Comment (TisonKun) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java]: We already have a similar util `ExecutionConfigAccessor. decodeUrlList`. I think we can reduce duplicates.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+		return ConfigUtils.decodeListFromConfig(configuration, configOption, url -> {
+			try {
+				return new URL(url);
+			} catch (MalformedURLException e) {
+				throw new IllegalArgumentException("Invalid URL", e);
+			}
+		});
+	}
+
+	private static Optional<JobID> getJobId(final Configuration configuration) {
```
Comment (TisonKun) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java]: It seems a poor util that we can inline.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/DetachedApplicationRunner.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+		final List<JobID> applicationJobIds = new ArrayList<>();
+		final PipelineExecutorServiceLoader executorServiceLoader =
+				new EmbeddedExecutorServiceLoader(applicationJobIds, dispatcherGateway);
+
+		try {
+			ClientUtils.executeProgram(executorServiceLoader, configuration, program);
+		} catch (ProgramInvocationException e) {
+			LOG.warn("Could not execute application: ", e);
+			throw new CompletionException("Could not execute application.", e);
```
Comment (TisonKun) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/DetachedApplicationRunner.java]: Use `FlinkRuntimeException` or handle in its `supplyAsync` caller site. We don't tight this method to run in future context. do we?

Comment (kl0u) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/DetachedApplicationRunner.java]: Maybe in a PR that I will open in the future, related to FLIP-85, but it is true that it is not needed here.

**Comment by kl0u** [comment]
Thanks for the review @TisonKun . I integrated your comments, let me know what you think.

**Comment by wangyang0918** [comment]
@kl0u Thanks for starting this great work. I started a standalone on K8s and submit some jobs via web dashboard. It works really well.

I just have one side comments. After introducing this changes, the users will not be able to submit the job that use `collect, count, etc.`. Is it the same behavior before this change?

**Comment by kl0u** [comment]
Thanks for looking into it @wangyang0918 ! Even before, batch jobs with `collect()` or `count()` were submitted but nothing was printed or could be done with the result, because we were extracting the `JobGraph` and submit it, without further executing the user's `main()`. So the only difference is that here we make it explicit.

**Comment by wangyang0918** [comment]
@kl0u Thanks for your reply. I have checked the current master behavior. The only difference is after this PR the WebUI will show an internal error. Since we make the `collect, count, etc.` more explicit. And we could find the exception in the jobmanager logs.
I think the behavior after this PR is just what i want.

**Comment by kl0u** [comment]
Super! Thanks @wangyang0918 ! @TisonKun do you have any additional comments on the PR?

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandler.java [hunk 5]
```diff
@@ -67,41 +69,39 @@ public JarRunHandler(
...
+
+		final PackagedProgram program = context.toPackagedProgram(configuration);
+
+		return CompletableFuture
+				.supplyAsync(() -> applicationRunner.run(gateway, program, configuration), executor)
+				.thenApply(jobIds -> {
+					if (jobIds.isEmpty()) {
+						throw new CompletionException(new ProgramInvocationException("No jobs submitted."));
+					}
+					return new JarRunResponseBody(jobIds.get(0));
```
Comment (TisonKun) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandler.java:103]: It is kind of subtle that we possibly actually have multiple job executed but return id of the first one. Maybe we should mention this manner somewhere.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+					}
+				});
+	}
+
+	@Override
+	public CompletableFuture<JobExecutionResult> getJobExecutionResult(final ClassLoader userClassloader) {
+		checkNotNull(userClassloader);
+
+		return dispatcherGateway
+				.requestJobResult(jobId, RpcUtils.INF_TIMEOUT)
```
Comment (TisonKun) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java:144]: I know the background that when we later in application mode scenario we have to retry and kill the application on terminate. But I think it is wried for others who will ask why `RpcUtils.INF_TIMEOUT` here.

Also, even we set to `RpcUtils.INF_TIMEOUT` it is actually about 9 months and possibly still we need retry on retiable exception(AskTimeout).

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandler.java (hunk 5 above)
Comment (kl0u) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandler.java:103]: Good point @TisonKun . 

I agree, that now it is possible to have multiple jobs being executed, but previously this was not possible because we were throwing an exception at the first `execute()` in order to extract the `JobGraph`. This is the reason I left it like this here. 

I could either document this behaviour here, or change the `JarRunResponseBody` to include multiple `JobIds` (which I find more consistent).

I also include @aljoscha  in the discussion to have more opinions.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java (hunk 6 above)
Comment (kl0u) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java:144]: You mean that we should go with a solution that retries periodically if the result is not there?

**Comment by kl0u** [comment]
I integrated your comments. Please have a look.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java (hunk 6 above)
Comment (kl0u) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java:144]: @TisonKun given that we are using directly the `dispatcher`, I am not so sure even if the `timeout` plays a role.

Comment (TisonKun) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java]: Every gateway instance is a proxy so I think `timeout` always works.

>You mean that we should go with a solution that retries periodically if the result is not there?

Yes if we meet a retriable exception.

For another perspective, if this method call by user, why not use configured `timeout`; if this method call by framework, you can take a look at `RestClusterClient#requestJobResult` where there is a retry mechanism underline.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/DetachedApplicationRunner.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * An {@link ApplicationRunner} which runs the user specified application
+ * using the {@link EmbeddedExecutor}.
+ * For this to work, this runner has to run on the same machine as the {@link DispatcherGateway}
```
Comment (aljoscha) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/DetachedApplicationRunner.java:46]: Wouldn't this also work if the dispatcher is on a different machine?

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.concurrent.CompletionException;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A {@link JobClient} with the ability to also submit jobs which
+ * uses directly the {@link DispatcherGateway}.
+ */
+@Internal
+public class EmbeddedClient implements JobClient {
```
Comment (aljoscha) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java:54]: Maybe `EmbeddedJobClient`? There's to many clients flying around already... 😅

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A {@link PipelineExecutor} that is expected to be executed on the same machine as the
+ * {@link org.apache.flink.runtime.dispatcher.DispatcherGateway Dispatcher} and uses it directly
```
Comment (aljoscha) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java:42]: Same here, this should technically work if the dispatcher is on a different machine, right? There's a couple of comments like this in the PR.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+	private static final Logger LOG = LoggerFactory.getLogger(EmbeddedExecutor.class);
+
+	public static final String NAME = "embedded";
+
+	private final Collection<JobID> applicationJobIds;
+
+	private final DispatcherGateway dispatcherGateway;
+
+	public EmbeddedExecutor(
```
Comment (aljoscha) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java:56]: We should clearly document that the `applicationJobIds` parameter is being filled by the executor, that this is basically for "returning" the job IDs. Also,  this could be called `submittedJobIds` here, because this knows nothing about applications.

Comment (aljoscha) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java:56]: This might also seem like a good case for an executor listener, that could be used to listen to the submitted job ids, but that's maybe a future change.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutorFactory.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+ * An {@link PipelineExecutorFactory} for the {@link EmbeddedExecutor}.
+ */
+@Internal
+public class EmbeddedExecutorFactory implements PipelineExecutorFactory {
+
+	private final Collection<JobID> applicationJobIds;
+
+	private final DispatcherGateway dispatcherGateway;
+
+	public EmbeddedExecutorFactory(
```
Comment (aljoscha) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutorFactory.java:42]: Same here for the job IDs. See comment on `EmbeddedExecutor`.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutorServiceLoader.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+	/**
+	 * Creates a {@link EmbeddedExecutorServiceLoader}.
+	 *
+	 * @param applicationJobIds a list initialized with the recovered jobs (if any). This list is also going to be filled
+	 *                          by the {@link EmbeddedExecutor} with the job ids of the new jobs that will be submitted. The
+	 *                          list can later be used for monitoring the status of the submitted/recovered jobs.
+	 * @param dispatcherGateway the dispatcher of the cluster which is going to be used to submit jobs.
+	 */
+	public EmbeddedExecutorServiceLoader(
+			final Collection<JobID> applicationJobIds,
```
Comment (aljoscha) [inline flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutorServiceLoader.java:54]: If this is meant to be used for two-way communication, maybe it's best to have two parameters. Otherwise things can be a bit confusing.

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java [hunk 13]
```diff
@@ -115,7 +120,15 @@ static void init() throws Exception {
 				LAST_SUBMITTED_JOB_GRAPH_REFERENCE.set(jobGraph);
 				return CompletableFuture.completedFuture(Acknowledge.get());
 			})
-			.build();
+			.setRequestJobResultFunction(jobID ->
```
Comment (aljoscha) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java:123]: Why this change?

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java [hunk 14]
```diff
@@ -63,7 +65,10 @@
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 
-/** Base test class for jar request handlers. */
+/**
+ * Base test class for jar request handlers.
+ * STILL THINK ABOUT THE ARGS AND THE REMAINING THINGS WE PUT IN THE PACKAGED PROGRAM.
```
Comment (aljoscha) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java:70]: ?

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerTest.java [hunk 15]
```diff
@@ -49,22 +49,12 @@
 	@ClassRule
 	public static final TemporaryFolder TMP = new TemporaryFolder();
 
-	enum Type {
```
Comment (aljoscha) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerTest.java:52]: Why these changes?

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/utils/JarHandlerUtils.java [hunk 16]
```diff
@@ -112,20 +117,43 @@ private JarHandlerContext(Path jarFile, String entryClass, List<String> programA
 			return new JarHandlerContext(jarFile, entryClass, programArgs, parallelism, jobId);
 		}
 
+		public void applyOnConfiguration(final Configuration configuration) {
```
Comment (aljoscha) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/utils/JarHandlerUtils.java:120]: `appleToConfiguration()`?

And maybe some Javadoc?

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandler.java [hunk 17]
```diff
@@ -67,41 +69,40 @@ public JarRunHandler(
...
-			final Executor executor) {
+			final Executor executor,
+			final Supplier<ApplicationRunner> applicationRunnerSupplier) {
 		super(leaderRetriever, timeout, responseHeaders, messageHeaders);
 
 		this.jarDir = requireNonNull(jarDir);
 		this.configuration = requireNonNull(configuration);
 		this.executor = requireNonNull(executor);
+
+		this.applicationRunner = applicationRunnerSupplier.get();
```
Comment (aljoscha) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandler.java:80]: Why is it a supplied if we immediately call it here. Why not call it later?

File: flink-core/src/main/java/org/apache/flink/configuration/PipelineOptionsInternal.java [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+import static org.apache.flink.configuration.ConfigOptions.key;
+
+/**
+ * Pipeline options that are not meant to be used by the user.
+ */
+@Internal
+public class PipelineOptionsInternal {
+
+	public static final ConfigOption<String> PIPELINE_FIXED_JOB_ID =
```
Comment (aljoscha) [inline flink-core/src/main/java/org/apache/flink/configuration/PipelineOptionsInternal.java:31]: We could also make this a regular public option but describe that it should only be used in special cases. Users will have to use it in the standalone case, where it would be good to also allow configuration of the fixed job id via the flink conf. WDYT?

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java (hunk 10 above)
Comment (kl0u) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutor.java:56]: I agree!

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java (hunk 14 above)
Comment (kl0u) [reply flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java:70]: Leftover :(

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutorServiceLoader.java (hunk 12 above)
Comment (kl0u) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/executors/EmbeddedExecutorServiceLoader.java:54]: What do you mean by "two-way communication"?

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java (hunk 13 above)
Comment (kl0u) [reply flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/handlers/JarHandlerParameterTest.java:123]: Removed. Leftover from previous version.

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandler.java (hunk 17 above)
Comment (kl0u) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarRunHandler.java:80]: For testing.

File: flink-core/src/main/java/org/apache/flink/configuration/PipelineOptionsInternal.java (hunk 18 above)
Comment (kl0u) [reply flink-core/src/main/java/org/apache/flink/configuration/PipelineOptionsInternal.java:31]: I would prefer to do it in a separate commit/jira because:
1) it seems an important change and
2) what happens in the case of multi-`execute()` applications? All the jobs will have the same id? And if we make it a list, what happens in the case of a non-deterministic `main()`.

File: flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java (hunk 6 above)
Comment (kl0u) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java]: I think you are right @TisonKun . I will update this but given that for now we are not using this method and we do not want to block a thread if the user does wait for the result by using explicitly the `JobClient`, I change the PR to return a special client which does not allow to fetch the result.

I will integrate the retry logic in a follow-up PR about the `StandaloneJobClusterEntrypoint`.

Comment (TisonKun) [reply flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedClient.java]: That should make sense.
